design/baremetalhost-v1alpha2.md

```diff
@@ -0,0 +1,236 @@
+access to the `go` library for parsing the string.
+
+The CPU speed is simply being reported, and not used by the
+`baremetal-operator` code. We could, therefore, use a `string` field
+without losing any functionality. The value we get from `gophercloud`
+is already a string.
+
+Since the fractional portion of the speed can be dropped, we can use
+an `int32` to hold the value. An integer value will make it easier for
+the `hardware-classification-controller` to continue to use the value
+when applying rules.
+
+The best approach is to use an `int32` field.
+
+### Converting Existing Data
+
+It is possible to [use a web
+hook](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#configure-customresourcedefinition-to-use-conversion-webhooks)
+to convert resources from their storage format to a specific API
+format. We do not want users to continue to consume the `v1alpha1`
+format, however, because we want to be able to drop support
+quickly. Therefore, we should not write a conversion web hook.
+
+Instead, we should add startup code to the baremetal-operator to read
+all host resources in their current format and write them (with type
+conversion but no other changes) to the new storage format. See option
+2
+[here](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#upgrade-existing-objects-to-a-new-stored-version)
+for more discussion.
+
+### Dropping Hardware Profile Field
+
+Now that we have support for more verbose root device hints and we
+have the `hardware-classification-controller`, we no longer need the
+hardware profile fields. We can remove them as part of the
+backwards-incompatible changes as we move to the new API version.
+
+### Risks and Mitigations
+
+We do not know all of the ways that consumers of the API are using the
+CPU speed values currently being reported. Given that most kubernetes
+tools appear to assume that floating point types are not to be used in
+APIs, it is likely that most consumers are either ignoring the value
+or using it in simple ways (displaying it, logging it, etc.) where
+loss of precision from the type conversion to a float is not
+significant.
+
+### Work Items
+
+- Release `baremetal-operator` version `0.1.0` with existing API.
+- Add new `v1alpha2` API
+- Change the `ClockSpeed` type to `int32` in `v1alpha2`.
+- Remove hardware profile from spec and status.
+- Remove the state machine states for setting the hardware profile.
+- Set `v1alpha2` as the storage version
+- Add conversion code to the `baremetal-operator`
+- Update the inspection code to no longer convert the value given by
+  Ironic Inspector before storing it.
+- Release `baremetal-operator` version `0.2.0` with code that can
+  upgrade existing resources.
+- Update the module references in `cluster-api-provider-metal3` and
+  `hardware-classification-controller`.
```

Comment:
Those will require a release. Considering that BMO is now deployed as part of CAPM3 for CAPI, we must release a new version of CAPM3 (that might only include the new BMO). We will need to make sure that the consumers do not skip that version (if the next release without v1alpha1 is already out) as it would break the deployment.

Reply:
Good points. I've added a comment about releasing the consumers. I'm not sure what to do about ensuring that users don't skip releases on an upgrade. Is a release note somewhere sufficient?